core/src/main/java/com/oath/oak/ThreadIndexCalculator.java

```diff
@@ -1,41 +1,62 @@
 
 package com.oath.oak;
 
-import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
 
 public class ThreadIndexCalculator {
 
     public static final int MAX_THREADS = 32;
     private static final int INVALID_THREAD_ID = -1;
-    private ThreadLocal<Integer> local = ThreadLocal.withInitial(() -> INVALID_THREAD_ID);
-    private AtomicInteger[] indices = new AtomicInteger[MAX_THREADS];
+    // Long for correctness and anti false-sharing
+    private AtomicLong[] indices = new AtomicLong[MAX_THREADS];
 
     private ThreadIndexCalculator() {
         for (int i = 0; i < MAX_THREADS; ++i) {
-            indices[i] = new AtomicInteger(INVALID_THREAD_ID);
+            indices[i] = new AtomicLong(INVALID_THREAD_ID);
         }
     }
 
+    private int getExistingIndex(long threadID){
+      int iterationCnt = 0;
+      int currentIndex = ((int)threadID) % MAX_THREADS;
+      long currentThreadID = indices[currentIndex].get();
+      while (currentThreadID != threadID) {
+        if (currentThreadID == INVALID_THREAD_ID) {
+          return -1*currentIndex;
```

Comment:
what does a negative number indicates?

Reply:
negative output indicates that a new index need to be created for this thread id
a comment is added